### ISSUE TITLE
null returned for 204

### ISSUE THREAD

**Issue by CanD42** [tiangolo/fastapi#449]
Hi,

we're running Python 3.7.3 with FastAPI 0.27.2 and have the following API definition (simplified):

```python
app.post("/url", status_code = HTTP_204_OK)
async def set_data(data: Data) -> None: 
   await db.set_data(data)
```

We return a 204 and return nothing in the body but we see that *null* is returned with a content length of 4. The client expect content-length = 0 for a 204 response. Can this please be fixed?

Thanks,
Daniel

**Comment by euri10** [comment]
FastAPI by default returns a JSONResponse and since this commit if I'm correct https://github.com/encode/starlette/commit/1654a48fa4f6edf5a4fc5dab118ef53ae0ef2ead Starlette return null for a empty JSONResponse

you can easily fix with 
`return Response(status_code=HTTP_204_NO_CONTENT)`

**Comment by dmontagu** [comment]
To add to @euri10 's point, you can similarly fix this via:

```python
from starlette.responses import Response

@app.post("/url", status_code = HTTP_204_OK, response_class=Response)
async def set_data(data: Data) -> None: 
   await db.set_data(data)
```

**Comment by CanD42** [comment]
Still baffled about the null return, but workaround works so I'll close the ticket.

Thanks

**Comment by euri10** [comment]
> Still baffled about the null return, but workaround works so I'll close the ticket.
> 
> Thanks

I agree, maybe worth a ping upstream as the current API let you return a 204 with content, which shouldn't be possible

**Comment by tiangolo** [comment]
Thanks for the help @euri10 and @dmontagu !

@CanD42 the reason for this is because, by default, FastAPI takes whatever you return and puts it in a JSON response. With all the headers, converting the data to valid JSON, etc.

And in Python, any function that doesn't return something, *implicitly* returns a `None`. And the equivalent of `None` in JSON is `null`. That's a special case where it isn't very intuitive but is actually because of the way the language works. And the fact that FastAPI takes automatically whatever "returned" by the function and puts it in JSON, instead of requiring you to return the instance of a `Response` class or similar.

Thanks for reporting back and closing the issue!

**Comment by iacopo-papalini** [comment]
Nevertheless, this is a bug that should be fixed because a 204 with a non-empty body is in violation of the HTTP RFC.

https://tools.ietf.org/html/rfc2616#section-10.2.5
` The 204 response MUST NOT include a message-body, and thus is always
   terminated by the first empty line after the header fields.`

**Comment by tiangolo** [comment]
For completeness, this was fixed in recent versions (I don't remember which).

This example:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/", status_code=204)
def main():
    print("done")
```

...doesn't return a body.

**Comment by remidinishanth-ntnx** [comment]
Thank you @tiangolo

**Comment by Kludex** [comment]
> For completeness, this was fixed in recent versions (I don't remember which).
 
 The content-length is still 4.
 
 > A server MUST NOT send a Content-Length header field in any response
   with a status code of 1xx (Informational) or 204 (No Content).  A
   server MUST NOT send a Content-Length header field in any 2xx
   (Successful) response to a CONNECT request (Section 4.3.6 of
   [RFC7231]).
   
I'll try to open a PR to fix this.

**Comment by owenlamont** [comment]
I just ran into the content-length 4 problem @Kludex - did you end up working on a fix?

**Comment by Kludex** [comment]
Either:
- Use `response_class=Response` on the decorator.
- Use `return Response(status_code=204)` on the function itself.

**Comment by Kludex** [comment]
I consider this as a bug, so I don't think it should be documented. I think 204 should not return a body, even if the response class is the default (`JSONResponse`), and follow the RFC mentioned above.

**Comment by TBBle** [comment]
Since this is still attracting comment after closing. I'll note that #2833 should resolve the need for `response_class`, by automating it when `status_code` is 204. And despite https://github.com/tiangolo/fastapi/issues/449#issuecomment-761838535, without the `response_class` I was seeing the 'null' body with FastAPI 0.70.0, so maybe the fix was reverted in the meantime.

There's more discussion in #2832 if you want to continue this on a non-closed bug, as I suspect this one's slipped off people's radars.
